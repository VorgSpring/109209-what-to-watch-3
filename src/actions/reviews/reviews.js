import {
  LOAD_REVIEWS_REQUEST,
  LOAD_REVIEWS_SUCCESS,
  LOAD_REVIEWS_ERROR
} from '../../constants/actions-type';

export const loadReviewsRequest = () => ({
  type: LOAD_REVIEWS_REQUEST
});

export const loadReviewsSuccess = (film) => ({
  type: LOAD_REVIEWS_SUCCESS,
  payload: film
});

export const loadReviewsError = (error) => ({
  type: LOAD_REVIEWS_ERROR,
  payload: error
});

