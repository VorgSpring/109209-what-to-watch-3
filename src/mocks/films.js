export const films = [
  {
    id: 1,
    bgImage: `/bg-image`,
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Crime`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah-blah`
    ],
    runTime: 123
  },
  {
    id: 2,
    bgImage: `/bg-image`,
    name: `Bohemian Rhapsody`,
    poster: `img/bohemian-rhapsody.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Romance`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 3,
    bgImage: `/bg-image`,
    name: `Macbeth`,
    poster: `img/macbeth.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Horror`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 4,
    bgImage: `/bg-image`,
    name: `Aviator`,
    poster: `img/aviator.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Dramas`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ]
  },
  {
    id: 5,
    bgImage: `/bg-image`,
    bgColor: `#fff`,
    name: `We need to talk about Kevin`,
    poster: `img/we-need-to-talk-about-kevin.jpg`,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Documentary`,
    released: 2000,
    preview: [
      {
        src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
        type: `video/mp4`
      },
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        type: `video/webm`
      }
    ],
    isFavorite: false,
    rating: 1,
    director: `blah`,
    scoresCount: 3,
    description: `blah`,
    starring: [
      `blah`, `blah`
    ],
    runTime: 123
  },
];

/* eslint-disable camelcase */
export const filmByServer = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  poster_image: `img/the-grand-budapest-hotel-poster.jpg`,
  preview_image: `img/the-grand-budapest-hotel.jpg`,
  background_image: `img/the-grand-budapest-hotel-bg.jpg`,
  background_color: `#ffffff`,
  video_link: `https://some-link`,
  preview_video_link: `https://some-link`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  rating: 8.9,
  scores_count: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  run_time: 99,
  genre: `Comedy`,
  released: 2014,
  is_favorite: false,
};
