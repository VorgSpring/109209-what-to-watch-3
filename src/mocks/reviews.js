import {review} from './review';

export const reviews = {
  odd: [
    review
  ],
  even: [
    review
  ]
};
