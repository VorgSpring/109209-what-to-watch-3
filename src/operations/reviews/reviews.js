import {
  loadReviewsRequest,
  loadReviewsSuccess,
  loadReviewsError
} from '../../actions/reviews/reviews';
import {ApiPaths} from '../../constants/api';

export const loadReviews = (filmId) => (dispatch, _getState, api) => {
  dispatch(loadReviewsRequest());

  return api.get(ApiPaths.getReviews(filmId))
    .then((response) => {
      const review = response.data;
      dispatch(loadReviewsSuccess({filmId, review}));
    })
    .catch((e) => {
      const {error} = e.response.data;
      dispatch(loadReviewsError({filmId, error}));
    });
};
