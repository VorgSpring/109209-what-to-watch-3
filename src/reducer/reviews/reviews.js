import {
  LOAD_REVIEWS_REQUEST,
  LOAD_REVIEWS_SUCCESS,
  LOAD_REVIEWS_ERROR,
  SEND_REVIEW_SUCCESS,
} from '../../constants/actions-type';

const initState = {
  isLoading: false,
  isError: false,
  reviews: null
};

export default (state = initState, action) => {
  switch (action.type) {
    case (LOAD_REVIEWS_REQUEST):
      return Object.assign(
          {}, state, {
            isLoading: true,
            isError: false
          }
      );

    // TODO сделать кеш не больше 3-х
    case (LOAD_REVIEWS_SUCCESS):
      return Object.assign(
          {}, state, {
            isLoading: false,
            reviews: Object.assign(
                {}, state.reviews, {
                  [action.payload.filmId]: action.payload.reviews
                }
            )
          }
      );

    case (LOAD_REVIEWS_ERROR):
      return Object.assign(
          {}, state, {
            isLoading: false,
            isError: true,
            reviews: Object.assign(
                {}, state.reviews, {
                  [action.payload.filmId]: []
                }
            )
          }
      );

    case (SEND_REVIEW_SUCCESS):
      return Object.assign(
          {}, state, {
            reviews: Object.assign(
                {}, state.reviews, {
                  [action.payload.filmId]: action.payload.reviews
                }
            )
          }
      );

    default:
      return state;
  }
};
