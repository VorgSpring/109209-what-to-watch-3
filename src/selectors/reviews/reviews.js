export const getReviewsByFilmIdSelector = (state, props) => {
  const reviews = state.reviews.reviews &&
    state.reviews.reviews[props.film.id];

  if (!reviews) {
    return null;
  }

  return reviews.reduce((acc, review, i) => {
    if (i % 2) {
      acc.odd.push(review);
    } else {
      acc.even.push(review);
    }
    return acc;
  }, {
    odd: [],
    even: []
  });
};
