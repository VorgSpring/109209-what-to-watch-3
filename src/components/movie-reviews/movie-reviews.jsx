import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {MovieReview} from '../movie-review/movie-review.jsx';
import {loadReviews} from '../../operations/reviews/reviews';
import {getReviewsByFilmIdSelector} from '../../selectors/reviews/reviews';

// TODO отрисовать пустые комментарии
export class MovieReviews extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {
      reviews,
      filmId,
      onLoadReviews
    } = this.props;

    if (!reviews) {
      onLoadReviews(filmId);
    }
  }

  componentDidUpdate({filmId}) {
    if (this.props.filmId !== filmId && !this.props.reviews) {
      this.props.onLoadReviews(this.props.filmId);
    }
  }

  render() {
    const {reviews} = this.props;

    return (
      <div className='movie-card__reviews movie-card__row'>
        <div className='movie-card__reviews-col'>
          {reviews && reviews.map((review) => (
            <MovieReview key={review.id} {...review} />
          ))}
        </div>
      </div>
    );
  }
}

MovieReviews.defaultProps = {
  reviews: null
};

MovieReviews.propTypes = {
  filmId: PropTypes.number.isRequired,
  reviews: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        user: PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired
        }).isRequired,
        rating: PropTypes.number.isRequired,
        comment: PropTypes.string.isRequired,
        date: PropTypes.string.isRequired
      })
  ),
  onLoadReviews: PropTypes.func.isRequired
};

const mapStateToProps = (state, props) => ({
  reviews: getReviewsByFilmIdSelector(state, props),
  filmId: props.film.id
});

const mapDispatchToProps = (dispatch) => ({
  onLoadReviews: (id) => dispatch(loadReviews(id))
});

export const MovieReviewsContainer = connect(
    mapStateToProps, mapDispatchToProps
)(MovieReviews);
