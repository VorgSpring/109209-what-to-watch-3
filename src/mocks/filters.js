import {ALL_GENRES_TYPE} from '../constants/films';

export const filters = {
  genre: ALL_GENRES_TYPE.ALL
};
