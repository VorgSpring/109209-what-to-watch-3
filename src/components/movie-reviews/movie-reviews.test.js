import React from 'react';
import {MovieReviews} from './movie-reviews.jsx';
import renderer from 'react-test-renderer';
import {reviews} from '../../mocks/reviews';

jest.mock(`../movie-review/movie-review.jsx`, () => ({
  MovieReview() {
    return <movie-review />;
  }
}));

describe(`MovieReviews`, () => {
  const onLoadReviews = () => {};

  it(`should renders correctly`, () => {
    const tree = renderer.create(
        <MovieReviews
          filmId={1}
          reviews={reviews}
          onLoadReviews={onLoadReviews}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should renders correctly without reviews`, () => {
    const tree = renderer.create(
        <MovieReviews
          filmId={1}
          onLoadReviews={onLoadReviews}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
