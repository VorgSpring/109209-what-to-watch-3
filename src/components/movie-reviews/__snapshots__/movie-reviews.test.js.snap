// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieReviews should renders correctly 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <movie-review />
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <movie-review />
  </div>
</div>
`;

exports[`MovieReviews should renders correctly without reviews 1`] = `
<div
  className="movie-card__reviews movie-card__row"
/>
`;
